docs/source/notes/numpy_compatibility.rst

```diff
@@ -0,0 +1,221 @@
+i.e. only leaf nodes are supported.
+
+.. code-block:: python
+
+   import torch
+   b_torch = torch.ones(3)
+   assert(b_torch.requires_grad == False)
+   assert(b_torch.is_cuda == False)
+   b_torch.numpy() # Fine, will work
+
+The interoperability guidelines places a strong emphasis on correctness over
+convenience. As the conversion to a NumPy array may potentially be a lossy
+operation (the computational graph), there are no inbuilt operations to do
+something which unconditionally returns a NumPy array.
+
+.. code-block:: python
+
+   # Never do this
+   def uncond_numpy(torch_tensor):
+        if torch_tensor.is_cuda:
+           warnings.warn("Moving back to CPU")
+           torch_tensor = torch_tensor.cpu()
+        if torch_tensor.requires_grad == False:
+            return torch_tensor.numpy()
+        else:
+            warnings.warn("Calling detach, probably lossy!")
+            return torch_tensor.detach().numpy()
+
+Operations
+------------
+
+All ``torch`` operators will helpfully fail with a ``TypeError`` if called with
+``numpy`` arrays. However, for **numpy methods**, using a ``torch.Tensor`` with
+an ``np.ndarray`` will return a ``torch.Tensor`` without creating any additional
+copies.
+
+ - Due to the :meth:`torch.Tensor.__array__()` implementation, a
+   ``np.ndarray`` which shares memory with the ``torch.Tensor`` is used for the
+   operation.
+ - The return type functionality is defined by
+   :meth:`torch.Tensor.__array_wrap__()`, and calls ``torch.from_numpy()``
+   internally.
+
+.. code-block:: python
+
+   import torch
+   import numpy as np
+   a_np = np.ones(3) # dtype=float64
+   b_torch = torch.ones(3) # dtype=float32
+   torch.add(a_np, b_torch) # TypeError
+   b_torch + a_np # TypeError
+   a_np + b_torch # OK but awkward
+   c_tmp = np.add(a_np, b_torch) # OK
+   # c_tmp.dtype is torch.float64
+
+Since the ``torch.Tensor`` methods will not work with ``np.ndarray`` objects, to
+prevent the apparent asymmetry of ``a_np * b_torch`` passing while ``b_torch *
+a_np`` fails, the user would **do best** to use NumPy functions instead:
+
+.. csv-table::
+   :header: Operator, NumPy Function, Description
+
+   "``+``", "``np.add()``", "Addition"
+   "``-``", "``np.subtract()``", "Subtraction"
+   "``*``", "``np.multiply()``", "Multiplication"
+   "``/``", "``np.divide()``", "Division"
+
+Conversions
+-------------
+
+Only a small subset of data type (``dtype``) objects defined in NumPy have an
+equivalent in PyTorch, namely:
+
+.. csv-table:: $ indicates the sizes supported, e.g. ``uint8``
+   :header: ``np.dtype``, ``torch.dtype``, sizes
+
+    "``bool_``", "``bool``", "N/A"
+    "``uint$``", "``uint$``", ":math:`8`"
+    "``int$``", "``int$``", ":math:`8, 16, 32, 64`"
+    "``float$``", "``float$``", ":math:`16, 32, 64`"
+    "``complex$``", "``complex$``", ":math:`64, 128`"
+
+.. warning::
+
+   The promotion and casting rules `of NumPy`_ differ from those `of PyTorch`_.
```

Comment:
This warning is not all that useful unless you explain how. What does the reader need to know here?

Reply:
I think the only point here is that not all the types are supported or have equivalents, but yeah, it is spurious here.